File: individual/client/src/main/java/requests/ErrorPacket.java
package requests;

import fields.OpCode;

import java.nio.ByteBuffer;

public class ErrorPacket extends Packet {

    private final String message;
    private final short error;

    public ErrorPacket(short error, String message) {
        super(OpCode.ERROR);
        this.error = error;
        this.message = message;
    }

    @Override
    public int size() {
        return Character.BYTES + Character.BYTES + message.length() + 1;
    }

    @Override
    public byte[] serialize() {
        ByteBuffer buf = ByteBuffer.allocate(size());
        buf.putShort(getOpCode());
        buf.putShort(error);
        buf.put(message.getBytes());
        buf.put((byte)0);
        return buf.array();
    }
}

File: udp/src/main/kotlin/ru/hse/anstkras/tftp/packet/ErrorPacket.kt
package ru.hse.anstkras.tftp.packet

import java.nio.ByteBuffer

class ErrorPacket(private val errorCode: Int, private val errorMessage: ByteBuffer) :
    Packet {

    //    2 bytes     2 bytes      string    1 byte
    //    -----------------------------------------
    //    | Opcode |  ErrorCode |   ErrMsg   |   0  |
    //    -----------------------------------------
    override fun getBytesRepresentation(): ByteBuffer {
        val buffer = ByteBuffer.allocate(errorMessage.remaining() + 5)
        buffer.putShort(5)
        buffer.putShort(errorCode.toShort())
        buffer.put(errorMessage)
        buffer.put(0)
        return buffer
    }

    companion object : Parsable<Packet> {
        override fun parse(byteBuffer: ByteBuffer): Packet {
            val errorCode = byteBuffer.short
            val errorMessage = PacketParser.parseStringToByteBuffer(byteBuffer)
            return ErrorPacket(errorCode.toInt(), errorMessage)
        }
    }

    fun getErrorMessage(): String {
        return "ErrorCode: $errorCode, $errorMessage"
    }
}